enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,175 @@
+* As an user of a IPI OpenShift cluster, I want clusterIP services to be working with dual-stack.
+
+### Goals
+
+- OpenStack VMs are configured with one dual-stack interface.
+- The IPv6 Subnets can be configured with: Stateless, SLAAC or Stateful.
+- The addresses on the VM's interface are configured with the same address
+assigned by OpenStack to the Port.
+- Nodes are configure with both IPv4 and IPv6 addresses, without exposing
+additional addresses, like storage interfaces.
+- Pods are configured with both IPv4 and IPv6 addresses.
+
+### Non-Goals
+
+- Support for OpenStack endpoints configured with dual-stack.
+- Support for LoadBalancer type Services.
+- Installer (IPI) handles creation of the Network and dual-stack Subnet.
+
+## Proposal
+
+Customers will be able to specify the IPv4 and IPv6 Subnets
+meant to be used in the `install-config.yaml`.
+
+### Workflow Description
+
+1. The customers will configure a network (tenant or provider) with IPv4 and
+IPv6 subnets. The IPv6 address mode for the Subnet can be of any type:
+SLAAC, Stateless or Stateful.
+2. The customer will create API and Ingress Ports using the Network created in the
+previous step.
+3. The customer will write the `install-config.yaml` and add the subnet IDs to
+the new field `machinesSubnets`.
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain `ip_addr_mode=eui64`
+and `method=auto` in the connection for the ethernet interface. While when using stateful
+it has to enable `ip=dhcp,dhcpv6` in the kernel arguments.
+5. The `uuid` from the network has to be added in the worker MachineSet alongside with the
+subnets that were pre-filled for the machine-api-provider-openstack to create one unique Port
+with both addresses configured for the Server.
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
+7. The installer will validate the subnets and VIPs, add the security group to the api and ingress
+ports and link the floating IPs to those Ports. Also, create the additional IPv6 security group rules
+and configured the servers with dual-stack addresses and add the dual-stack allowed address pairs.
+8. In machine-config-operator, kubelet will be update to not be configured with node-ips.
+9. cloud-provider-openStack will filter the node addresses that are meant to be specified in the Status of the Node
+by looking for a tag that was applied in the main Ports of the Node.
+
+### API Extensions
+
+The `machinesSubnet` field, which is available in the `install-config.yaml` under
+`platform` and `openstack` will be deprecated in favor of a new field named `machinesSubnets`
+to persist both IPv4 and IPv6 provided Subnets.
```

Comment:
I'm wondering if we shouldn't take the opportunity to deprecate `machinesSubnet` and just use Failure Domains `portTargets` and not do `machinesSubnets` because it seems we're achieving the same goals here: decide the port configuration for OCP masters.

cc @pierreprinetti @mdbooth 

Workers can be figured out on day 2.

Reply:
From a purely functional perspective, you can achieve with failure domains the same configuration as you get with `machinesSubnets` by assigning each node to a failure domain with the same `portTarget` with `id: control-plane` (that's what it's called, because failureDomains is masters-only at the moment). With one caveat: the "default" network.

I see some potential issues with the deprecation of `machinesSubnet(s)`:

* Currently, if you don't pass a `machinesSubnet`, the Installer creates the default network. If you pass a `machinesSubnet`, that network becomes the default network. How does the deprecation work? Do we stop creating a default network? If so, what happens if day-2 you create a failure domain that doesn't specify any `control-plane` `portTarget`?
* From the perspective of a user wanting to install dual-stack, who is not interested in failure domains. Would they have to use failure domains for the purpose of attaching two subnets to their machines?
* Last but not least, failure domains do not currently exist for workers.

Reply:
>What is the impact of the id:control-plane, where is it used?

In Failure domains, you can pass zero or more `portTargets`. Each `portTarget` is one additional network to attach to the machine. Each `portTarget` has a unique `id` (for reasons not related to this conversation).

As a special case, when the id of a `portTarget` is exactly `control-plane`, then that port is set to be the first in the Machine's `Networks`, and the machine does not get attached to the "default" port.

Reply:
@mdbooth Very good point. So, your suggestion would be in line with what @pierreprinetti proposed [here](https://github.com/openshift/enhancements/pull/1365#discussion_r1147472319) to use `portOpts`, right?
and probably that would be a separate field from failureDomain just how it's done at the moment for `machinesSubnet`, correct?

Reply:
Whatever direction we go with this, please can the enhancement include example install-config demonstrating whatever we're proposing?

Reply:
@mdbooth @pierreprinetti can you have another look at this topic if your thought has changed about using `portTarget` instead of `machineSubnets`. Here is how I was developing it https://github.com/openshift/installer/pull/6797/files#diff-2ce80af83d1ead3e93b6ef440f1ba555ef2fb4cf285bbf06a3323e9132ee6133R126